### PR TITLE
Distributed nearest callback

### PR DESCRIPTION
**Pull request by aprokop**
[The 3-round rework of #737 will be pushed in that branch once this PR has been merged].

Works for both APIv1 and APIv2 (that was quite tricky).

Introduces two ways to do a nearest callback in the distributed setting:
- 2-round communication for a constrained callback

Constrained meaning that
- It is allowed to be called on not final matches
- For each match, it outputs exactly one result

Constrained callback is indicated with a new tag.

Current limitations:
- I have not thought yet about post-callbacks and whether they should be allowed
  To be fair, even with the spatial callbacks I'm not sure what that means

### CODE REVIEW

File: src/details/ArborX_Callbacks.hpp [hunk 1]
```diff
@@ -44,6 +46,25 @@ struct DefaultCallback
...
 };
 
+#ifdef ARBORX_ENABLE_MPI
+struct ConstrainedNearestCallbackTag
+{};
+
+struct DefaultCallbackWithRank
+{
+  using tag = ConstrainedNearestCallbackTag;
+
```
Comment (aprokop) [inline src/details/ArborX_Callbacks.hpp:56]: @dalg24 I know you don't like this approach (similar to the discussion we had in #675). What alternative do you propose?

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  {
+    auto const ray = getGeometry(x.predicates(i));
+    return intersects(ray);
+  }
+};
+
+namespace Details
+{
+
+template <typename Tree, typename Callback, typename OutValue, bool UseValues>
```
Comment (aprokop) [inline src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:99]: The `OutValue` argument is really annoying. Currently, no avoiding it. The issue is that we can't determine from the callback's signature what the output value type is going to be. The way we are doing things right now is that determine that from the storage arrays passed in by a user.

File: src/details/ArborX_DetailsLegacy.hpp [hunk 3]
```diff
@@ -84,7 +84,7 @@ struct LegacyCallbackWrapper
                                   PairValueIndex<Value, Index> const &value,
                                   Output const &out) const
   {
-    _callback(predicate, value.index, out);
+    _callback(predicate, (int)value.index, out);
```
Comment (aprokop) [inline src/details/ArborX_DetailsLegacy.hpp:87]: This is a more precise treatment for APIv1. I also needed this due to the default distributed callback now being templated on the value, resulting in the `unsigned -> signed` loss of precision errors.

File: test/tstDistributedTree.cpp [hunk 4]
```diff
@@ -31,6 +31,41 @@ namespace tt = boost::test_tools;
...
+{
+  using tag = ArborX::Details::ConstrainedNearestCallbackTag;
+
+  int rank;
+
+  template <typename Predicate, typename Value, typename OutputFunctor>
+  KOKKOS_FUNCTION void operator()(Predicate const &, Value const &value,
+                                  OutputFunctor const &out) const
+  {
+    out({rank, value});
```
Comment (dalg24) [inline test/tstDistributedTree.cpp:65]: Why do we need both this one and `DefaultCallbackWithRank`?
One aggregates the rank first and the value second, the other swaps the two.

Comment (aprokop) [reply test/tstDistributedTree.cpp:65]: We don't *need* it. Rather, this is just a custom callback that does this.

File: src/details/ArborX_Callbacks.hpp (hunk 1 above)
Comment (dalg24) [reply src/details/ArborX_Callbacks.hpp:56]: You define a decorator that wraps the callback and you check whether the callback is decorated.

Something like
```C++
template <class Callback>
struct DecoratedCallback {
  Callback callback_;
  template <class... Args>
  void operator()(Args&&... args) const { callback_((Args&&)args...); }
};

template <class Callback>
struct is_decorated_callback : std::false_type {};

template <class Callback>
struct is_decorated_callback<DecorateCallback<Callback>> : std::true_type {};
```
Try to come up with more descriptive names.
I am not convinced this belongs to the general callback header, you might want to move it to the distributed headers.

Comment (aprokop) [reply src/details/ArborX_Callbacks.hpp]: Opted for the decorator instead of updating `operator()`.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+template <>
+struct is_constrained_callback<DefaultCallback> : std::true_type
+{};
+template <>
+struct is_constrained_callback<DefaultCallbackWithRank> : std::true_type
+{};
+
+template <class Callback>
+inline constexpr bool is_constrained_callback_v =
+    is_constrained_callback<Callback>::value;
```
Comment (aprokop) [inline src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:84]: @dalg24 Please look at this approach.

File: src/details/ArborX_DetailsLegacy.hpp [hunk 6]
```diff
@@ -96,24 +96,10 @@ struct LegacyDefaultCallback
                                   PairValueIndex<Value, Index> const &value,
                                   OutputFunctor const &output) const
   {
-    output(value.index);
+    output((int)value.index);
   }
 };
 
-#ifdef ARBORX_ENABLE_MPI
-struct LegacyDefaultCallbackWithRank
```
Comment (aprokop) [inline src/details/ArborX_DetailsLegacy.hpp:104]: I moved this to distributed headers and renamed. It no longer is being called only from APIv1, but also from within APIv2.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+struct is_constrained_callback : std::false_type
+{};
+template <class Callback>
+struct is_constrained_callback<
+    Experimental::ConstrainedDistributedNearestCallback<Callback>>
+    : std::true_type
+{};
+template <>
+struct is_constrained_callback<DefaultCallback> : std::true_type
+{};
```
Comment (dalg24) [inline src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:77]: It is non-obvious to me why we need a specialization for the default callback (haven't thought too hard about it) so it might be worth a comment.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp (hunk 5 above)
Comment (dalg24) [reply src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:84]: ok

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+    auto box = getGeometry(x.predicates(i));
+    auto &min_corner = box.minCorner();
+    auto &max_corner = box.maxCorner();
+    auto const distance = x.distances(i);
+    for (int d = 0; d < 3; ++d)
+    {
+      min_corner[d] -= distance;
+      max_corner[d] += distance;
+    }
+    return intersects(box);
```
Comment (dalg24) [inline src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:133]: This enlarging the box can yield false positive right?  Is that potentially an issue?
We might want to add a comment.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+    return intersects(Sphere{sphere.centroid(), distance + sphere.radius()});
+  }
+  template <
+      class Dummy = Geometry,
+      std::enable_if_t<std::is_same_v<Dummy, Geometry> &&
+                       std::is_same_v<Dummy, Experimental::Ray>> * = nullptr>
+  static KOKKOS_FUNCTION auto get(Self const &x, size_type i)
+  {
+    auto const ray = getGeometry(x.predicates(i));
+    return intersects(ray);
```
Comment (dalg24) [inline src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:151]: Remind me, don;t we only intersect with geometries in front of the origin?
In that case should we be moving the origin backwards for the predicate?

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+  {}
+
+  template <typename Query, typename Value, typename Output>
+  KOKKOS_FUNCTION void operator()(Query const &query, Value const &value,
+                                  Output const &out) const
+  {
+    if constexpr (UseValues)
+    {
+      OutValue out_value;
+      [[maybe_unused]] int count = 0;
```
Comment (dalg24) [inline src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:178]: Why did you add the `[[maybe_unused]]` attribute here?
It is being incremented below.  Were you actually seeing warning with `NDEBUG` defined?

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+      _callback(query, value, [&](OutValue const &ov) {
+        out_value = ov;
+        ++count;
```
Comment (dalg24) [inline src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:181]: Doesn't this incrementing the counter need to be atomic?

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+      });
+      KOKKOS_ASSERT(count == 1);
```
Comment (dalg24) [inline src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:183]: Why can it only be called once?  It might be worth a comment.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+  Kokkos::View<unsigned int *, typename Tree::memory_space> _rev_permute;
+
+  template <typename ExecutionSpace>
+  CallbackWithDistance(ExecutionSpace const &exec_space, Tree const &tree,
+                       Callback const &callback)
+      : _tree(tree)
+      , _callback(callback)
+  {
+    // NOTE cannot have extended __host__ __device__  lambda in constructor with
+    // NVCC
```
Comment (dalg24) [inline src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:216]:
```suggestion
    // NOTE cannot have extended __host__ __device__ lambda in constructor with
    // NVCC
```

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+  void computeReversePermutation(ExecutionSpace const &exec_space)
+  {
+    auto const n = _tree.size();
+
+    _rev_permute = Kokkos::View<unsigned int *, typename Tree::memory_space>(
+        Kokkos::view_alloc(
+            Kokkos::WithoutInitializing,
+            "ArborX::DistributedTree::query::nearest::reverse_permutation"),
+        n);
+    if (!_tree.empty())
```
Comment (dalg24) [inline src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:230]: I would prefer `if(tree.empty()) return;` first thing in this function

File: test/tstDistributedTree.cpp (hunk 4 above)
Comment (dalg24) [reply test/tstDistributedTree.cpp:63]: Maybe add a comment to that effect then.

File: src/ArborX_DistributedTree.hpp [hunk 15]
```diff
@@ -190,39 +190,14 @@ class DistributedTree<MemorySpace, Details::LegacyDefaultTemplateValue,
...
-      KokkosExt::reallocWithoutInitializing(space, indices_and_ranks, n);
-      Kokkos::parallel_for(
-          "ArborX::DistributedTree::query::zip_indices_and_ranks",
-          Kokkos::RangePolicy<ExecutionSpace>(space, 0, n),
-          KOKKOS_LAMBDA(int i) {
-            indices_and_ranks(i) = {values(i), ranks(i)};
-          });
-    }
+    int comm_rank = -1;
+    if (base_type::getComm() != MPI_COMM_NULL)
```
[the remark this answers is not in the document]
Comment (aprokop) [reply src/ArborX_DistributedTree.hpp:194]: Not for the default constructed one.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp (hunk 7 above)
Comment (aprokop) [reply src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:77]: We need it because we use default callback internally when implementing a distributed nearest query without user provided callback.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp (hunk 8 above)
Comment (aprokop) [reply src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:133]: The behavior is unchanged since `master`. It can do false positive for the upper tree, resulting in more work to be done. From a user perspective, there is no false positives.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp (hunk 9 above)
Comment (aprokop) [reply src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:151]: The behavior is unchanged from the `master`. Ray is special since the meaning of the nearest is different for it (nearest only among those intersected). Since the intersection of rank boxes would provide the correct ordering, this does not require a change.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp (hunk 10 above)
Comment (aprokop) [reply src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:178]: The compiler [complained](https://cloud.cees.ornl.gov/jenkins-ci/job/ArborX/job/PR-1075/2/gcc/folder.1897384215/) about it.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp (hunk 11 above)
Comment (aprokop) [reply src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:181]: This assumes that each query is executed by a single thread. If that assumption is wrong, we don't have an easy way to check this, unless we create a full view and do atomics.

File: test/tstDistributedTree.cpp (hunk 4 above)
Comment (aprokop) [reply test/tstDistributedTree.cpp:65]: Done.

File: src/details/ArborX_DetailsLegacy.hpp (hunk 3 above)
Comment (aprokop) [reply src/details/ArborX_DetailsLegacy.hpp:90]: Added a comment.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp (hunk 7 above)
Comment (aprokop) [reply src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:79]: Added a comment.

**Comment by aprokop** [comment]
@dalg24 Should have addressed all your comments with one exception.

Regarding atomics, I think it is fine as it is. I did not do atomics, as I think it is fine as it is. The only way I can see us triggering it if we call callbacks on the on-node nearest neighbor search results in parallel. We don't currently do that, nor do we have any plans to do it, as nearest neighbors require priority trees. The only exception to the latter is k = 1 specialization in BruteForce (not currently implemented), but even that would result in a single answer and thus no parallel callback calls for the same query.

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+#include <ArborX_Sphere.hpp>
+
+namespace ArborX
+{
+
+namespace Experimental
+{
+
+// Constrained callback is a callback that a user promises to:
+// - be not pure
```
Comment (dalg24) [inline src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:29]: I don't get the "not pure".  Plz explain.

**Review by dalg24 — COMMENTED** [review]
Just a few last comments

File: src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp (hunk 16 above)
Comment (dalg24) [reply src/details/ArborX_DetailsDistributedTreeNearestHelpers.hpp:29]: Please clarify that takes two arguments

File: src/details/ArborX_DetailsLegacy.hpp (hunk 3 above)
Comment (dalg24) [reply src/details/ArborX_DetailsLegacy.hpp:90]: Needed for aggregate initialization

**Review by dalg24 — APPROVED** [review]
Merge after you add the comments I asked for

**Comment by aprokop** [comment]
4 builds (all CUDA and SYCL) timed out, though 2 of them made through the first make+ctest. 3 builds passed. Given that there's been only comment change since 94330e2, I'm comfortable merging this.
